### ISSUE TITLE
A Neural Algorithm of Artistic Style

### ISSUE THREAD

**Issue by zachmayer** [dmlc/mxnet#664]
I was wondering if anyone would be interested in helping me replicate the images from this paper?
http://arxiv.org/abs/1508.06576

It looks like it's just a bunch of covnets, so we could possibly start with the pre-trained models and then try to reverse engineer the author's approach for combining images.

It'd make for a really, really cool mxnet demo =D

**Comment by mli** [comment]
that's absolutely awesome! we can also setup a website so others can try it easily

**Comment by zachmayer** [comment]
see also http://www.deepart.io/

**Comment by mli** [comment]
exactly. i think we can speedup the prediction time a lot comparing them:)

**Comment by zachmayer** [comment]
Agreed!  I also want to try it out locally =D

**Comment by mli** [comment]
sure. please ping me at anytime if you have problem to get it run

**Comment by tqchen** [comment]
For web-service, related to #629

**Comment by zachmayer** [comment]
@mli I haven't even started yet, so any suggestion you have would be appreciated.  I'm try to figure out how to begin:  should I start with a pre-trained network, or should I somehow train a network on a pair of images alone?

**Comment by mli** [comment]
i guess we can first reuse the pretrained caffe models https://github.com/jcjohnson/neural-style/tree/master/models (mxnet can load caffe models)

then try to translate their lua codes to mxnet

we may consider train the network later. we already have vgg model training codes there
https://github.com/dmlc/mxnet/tree/master/example/image-classification

**Comment by phunterlau** [comment]
Training the model needs much time and tune. I have tried the Lua version of  https://github.com/jcjohnson/neural-style/tree/master/models for producing neural style pictures, and it worked nicely. The lua version loads the caffe model, so mxnet can load the same model too from model.sh and simply replicate the producing script from Lua to python with mxnet.

![Pogo with oil painting style](https://cloud.githubusercontent.com/assets/1690881/11313823/ef26f228-8f94-11e5-8757-9e583a2b63d7.png)

**Comment by antinucleon** [comment]
The example looks cool. I think we can do it on Inception instead of VGG

**Comment by phunterlau** [comment]
is inception faster than vgg? vgg model is so slow as 2min on a titan x, which prevents it from being production

**Comment by winstywang** [comment]
I would like to do it, but the priority is not that high so far.

**Comment by pluskid** [comment]
@winstywang I'm recently taking a painting class, one of the project is to apply some specific artistic style to paint a photograph. If you implemented this, I should show it to the instructor! Haha! :laughing:

**Comment by antinucleon** [comment]
I am working on it...  First I will repeat torch experiment, then I will replace VGG.

**Comment by pluskid** [comment]
@antinucleon wondering why people use VGG so much despite it being much bigger and slower than google's model. The recently announced [neuraltalk2](https://github.com/karpathy/neuraltalk2) also uses VGG for example. Would be very interesting if we could show here that the huge VGG model could be replaced.

**Comment by antinucleon** [comment]
![download](https://cloud.githubusercontent.com/assets/1258974/11377043/3d0ea44a-92a1-11e5-9910-c5b3099f82e3.png)

I have done most parts, but seems there are minor issues need to be fixed.

**Comment by phunterlau** [comment]
superb, what is the speed? i can start drafting a blog of advertising it

**Comment by zachmayer** [comment]
Someone made a demo for tensorflow too: https://github.com/woodrush/neural-art-tf

**Comment by phunterlau** [comment]
Saw the tensor flow demo too. They use the exact same caffe model as in the Lua version. Seems like the number of iteration is not large enough for understanding the style. It should be around 800-100 from the Lua version, which is the reason why take so long (2 min with Titan X). If mxnet can simply this iteration for searching for optimal parameters, mxnet can save much time. The lua version mentioned their `-optimizer` parameter in the README https://github.com/jcjohnson/neural-style

**Comment by zachmayer** [comment]
I'm excited to see the results!  Also, there's already another one:
https://github.com/anishathalye/neural-style

tensor flow has some great marketing...

**Comment by mli** [comment]
![out16](https://cloud.githubusercontent.com/assets/421857/11384641/75843e5a-92df-11e5-9960-102346a3e95b.jpg)

we can almost reproduce the results, thanks to @antinucleon 

but the major problem is the speed. it needs thousands of iterations which takes minutes on a decent gpu cards..

**Comment by zachmayer** [comment]
@mli Is there code somewhere to reproduce the results I could try out myself?

**Comment by antinucleon** [comment]
@zachmayer I will commit it soon :)

**Comment by zachmayer** [comment]
@antinucleon Really looking forward to it!

**Comment by antinucleon** [comment]
https://github.com/dmlc/mxnet/tree/master/example/neural-style

**Comment by phunterlau** [comment]
so cool, how is the speed? let me write a blog for advertising it.

**Comment by mli** [comment]
not well done yet. @antinucleon is still on training a smaller model to make it fast

**Comment by zachmayer** [comment]
I tried the example, and got `Please compile with CUDA enabled`.  Is it silly to try to run this example without cuda?  I don't need a ton of speed and want to try it on a device without a gpu.

**Comment by mli** [comment]
use --gpu -1 to disable gpu

**Comment by phunterlau** [comment]
Just an estimation that the Lua version took about 40-60 minutes on CPU and 2 min on GPU, so we may expect 20-30 min on CPU with mxnet

**Comment by phunterlau** [comment]
just had some quick benchmark plus comparison with Lua version, as well as some tricks for better results.

1. Memory: with the same image size of 512px, Lua version takes 2868MB memory while mxnet takes 1498MB. It makes mxnet unique advantages that, laptop users can try neural art because most laptop has <2GB GPU memory, also mxnet can train larger pictures on my poor man's GTX 960 4 GB, where 800px can easily core-dump the Lua version.

2. speed: Lua version took 232 seconds on my poor man's GTX 960 (1000 iterations) while mxnet takes 93 seconds for converging into a similar result in 200 iterations. Not sure if there will be a simplified model for less time. Also, mxnet loads a much smaller model file, although it is the same caffe model.

3. tricks of having better outputs. the default denoise factor `remove-noise` is too large which makes the picture looks blur, a better value is about 0.15 instead of 0.2 . `stop-eps` value needs to be a little bit smaller for a stronger style, e.g. 0.004 or 0.003, but it may take longer time to converge. Set `content-weight` to 20 instead of 10 keeps more content shape, otherwise the style is so strong that distort the original content.

Here it is the same cat `pogo` with Van Gogh:
<img width="512" alt="pogo-vangogh-mxnet" src="https://cloud.githubusercontent.com/assets/1690881/11481189/c40ceb24-9750-11e5-8c7a-fc46c4ac675c.png">

**Comment by antinucleon** [comment]
@phunterlau Guess we can make a cool blog about your finding and post in reddit or other place. For cuDNN bug, I still don't have any idea yet.

**Comment by antinucleon** [comment]
If switch to CuDNN, we can make at least 40% faster. So think speed won't be an issue.

**Comment by phunterlau** [comment]
most people don't have CuDNN and they just want to have a try without pain from installing Lua, so for a blog post (if mxnet committee agrees for a blog, I can start drafting one), CuDNN is not needed. As to the ultimate solution, CuDNN is worth doing, considering a GTX 980 can reach 50s, so a Titan X with CuDNN can make it 20 s? It is an significant improvement.

If it is OK with a blog now, I can start writing.

**Comment by phunterlau** [comment]
For people who want a CPU only benchmark, here it is: on clang-omp CPU only with 2 cores from an i7 macbook pro, for having the same result, it takes 39 mins total. Will a faster learning rate on `lr` will help? current it is 0.1. I am going to experiment more with GPU anyway.

**Comment by mli** [comment]
@phunterlau i found the trick is to use a figure as large as possible. for example, i updated the sample output by change to `--max-long-edge 900`, it looks much better

**Comment by phunterlau** [comment]
@mli then how much memory does it take? I only have poor man 4gb

**Comment by mli** [comment]
3.7gb;)

**Comment by phunterlau** [comment]
sound like my poor GTX 960 problem: when I have max-long-edge greater than 700, it gives:

```
INFO:root:load the content image, size = (1280, 960)
INFO:root:resize the content image to (780, 585)
[22:42:54] ./dmlc-core/include/dmlc/logging.h:208: [22:42:54] src/operator/./convolution-inl.h:258: Check failed: (param_.workspace) >= (required_size)
Minimum workspace size: 1168128000 Bytes
Given: 1073741824 Bytes
[22:42:54] ./dmlc-core/include/dmlc/logging.h:208: [22:42:54] src/engine/./threaded_engine.h:295: [22:42:54] src/operator/./convolution-inl.h:258: Check failed: (param_.workspace) >= (required_size)
Minimum workspace size: 1168128000 Bytes
Given: 1073741824 Bytes
An fatal error occurred in asynchronous engine operation. If you do not know what caused this error, you can try set environment variable MXNET_ENGINE_TYPEto NaiveEngine and run with debugger (i.e. gdb). This will force all operations to be synchronous and backtrace will give you the series of calls that lead to this error. Remember to set MXNET_ENGINE_TYPE back to empty after debugging.
terminate called after throwing an instance of 'dmlc::Error'
  what():  [22:42:54] src/engine/./threaded_engine.h:295: [22:42:54] src/operator/./convolution-inl.h:258: Check failed: (param_.workspace) >= (required_size)
Minimum workspace size: 1168128000 Bytes
Given: 1073741824 Bytes
An fatal error occurred in asynchronous engine operation. If you do not know what caused this error, you can try set environment variable MXNET_ENGINE_TYPEto NaiveEngine and run with debugger (i.e. gdb). This will force all operations to be synchronous and backtrace will give you the series of calls that lead to this error. Remember to set MXNET_ENGINE_TYPE back to empty after debugging.
Command terminated by signal 6
```

**Comment by mli** [comment]
you can replace `workspace=1024` in model_vgg19.py into `workspace=2048`

**Comment by phunterlau** [comment]
That works. I can maximize the edge to 850 by changing workspace to 2048. Does it have special meanings?

![pogo-vangogh-30-2-0 003-0 3-out-850](https://cloud.githubusercontent.com/assets/1690881/11582168/e4e1caca-99ff-11e5-8a30-7090b1b8e1b7.jpg)

**Comment by mli** [comment]
it increases the temp buffer size, or the working space size

**Comment by mli** [comment]
i'm going to close this issue now since the demo is already available at https://github.com/dmlc/mxnet/tree/master/example/neural-style

further work is definitely required to improve both the speed and the results.
